File: lib/year2019/day13.rb
require 'io/console'
require 'solver'
require 'year2019/intcode_computer'

module Year2019
  class Day13 < Solver
    def part_one
      arcade_game.play_game.block_count
    end

    def part_two
      arcade_game(coins: 2).play_game.score
    end

    def arcade_game(coins: nil, interactive: false, display: false)
      ArcadeGame.new(program, coins, interactive, display)
    end

    def program
      @program ||= raw_input.split(",").map(&:to_i)
    end

    class ArcadeGame
      def initialize(program, coins, interactive, display)
        @computer = IntcodeComputer.new program
        @computer.set(0, coins) if coins
        @screen_hash = Hash.new
        @outputs = []
        @interactive = interactive
        @display = display
      end

      def inspect
        screen
      end

      def interactive?
        @interactive
      end

      def display?
        @display
      end

      def advance
        if @computer.requires_input?
          handle_input
        elsif @computer.will_output?
          handle_output
        else
          @computer.advance
        end
      end

      def play_game
        advance until @computer.halted?
        self
      end

      def block_count
        @screen_hash.values.count(2)
      end

      def score
        @score
      end

      def screen
        return "new ArcadeGame" if @screen_hash.empty?

        Grid.display(
          @screen_hash,
          :type => 'hash',
          "0" => " ",
          "1" => "|",
          "2" => "+",
          "3" => "_",
          "4" => "o",
        )
      end

      def display!
        return unless display? || interactive?
        system "clear"
        puts "\nscore: #{score}\n#{screen}\n"
      end

      def store_outputs
        *position, value = @outputs
        if position == [-1, 0]
          @score = value
        else
          @screen_hash[position] = value
        end
        @outputs = []
      end

      def process_outputs
        if @outputs.size == 3
          store_outputs
          display!
        end
      end

      def handle_output
        @outputs << @computer.next_output
        process_outputs
      end

      def handle_input
        @computer.next_input do |computer|
          computer.add_input input_value
        end
      end

      def input_value
        interactive? ? interactive_input : automatic_input
      end

      def interactive_input
        { 'h' => -1, 'l' => 1 }.fetch STDIN.getch, 0
      end

      def automatic_input
        if paddle < ball
          1
        elsif paddle > ball
          -1
        else
          0
        end
      end

      def paddle
        @screen_hash.key(3).first
      end

      def ball
        @screen_hash.key(4).first
      end
    end
  end
end

File: rust/src/year2019/day13.rs
use std::collections::HashSet;

use crate::{
    io::{Input, Solution},
    point::Point,
    year2019::computer::{Computer, Program, IO},
};

pub fn solve(part: &str, input: &Input) -> Solution {
    Solution::build(part, input, &part_one, &part_two)
}

fn part_one(input: &Input) -> i64 {
    block_count(input)
}

fn part_two(input: &Input) -> i64 {
    Arcade::from(input).play()
}

fn block_count(input: &Input) -> i64 {
    let mut computer = Computer::new(Program::from(input));
    let mut count = 0;
    while !computer.terminated() {
        computer.next_output();
        computer.next_output();
        if computer.next_output() == 2 {
            count += 1;
        }
    }
    count
}

struct Arcade {
    computer: Computer,
    paddle: Point,
    ball: Point,
    blocks: HashSet<Point>,
    score: i64,
}

impl Arcade {
    fn play(&mut self) -> i64 {
        while !self.computer.terminated() {
            self.step();
        }
        self.score
    }

    fn step(&mut self) {
        match self.computer.next_io() {
            IO::Terminated => {}
            IO::Input => self.handle_input(),
            IO::Output => self.handle_output(),
        }
    }

    fn handle_input(&mut self) {
        let input = match self.paddle.x.cmp(&self.ball.x) {
            std::cmp::Ordering::Less => 1,
            std::cmp::Ordering::Greater => -1,
            std::cmp::Ordering::Equal => 0,
        };
        self.computer.input(input);
    }

    fn handle_output(&mut self) {
        self.computer.interact();
        let x = *self.computer.output() as i32;
        let y = self.computer.next_output() as i32;
        let pt = Point::new(x, y);
        match (x, y, self.computer.next_output()) {
            (-1, 0, score) => {
                self.score = score;
            }
            (_, _, 2) => {
                self.blocks.insert(pt);
            }
            (_, _, 3) => {
                self.paddle = pt;
            }
            (_, _, 4) => {
                self.ball = pt;
            }
            _ => {}
        }
    }
}

impl From<&Input> for Arcade {
    fn from(input: &Input) -> Self {
        let mut program = Program::from(input);
        program.set(0, 2);
        let computer = Computer::new(program);
        let paddle = Point::new(0, 0);
        let ball = Point::new(0, 0);
        let blocks = HashSet::new();
        let score = 0;
        Self {
            computer,
            paddle,
            ball,
            blocks,
            score,
        }
    }
}
